Python/optimizer_analysis.c

```diff
@@ -83,6 +1101,93 @@ remove_unneeded_uops(_PyUOpInstruction *buffer, int buffer_size)
     }
 }
 
+static inline bool
+op_is_zappable(int opcode)
+{
+    switch(opcode) {
+        case _SET_IP:
+        case _CHECK_VALIDITY:
+        case _LOAD_CONST_INLINE:
+        case _LOAD_CONST:
+        case _LOAD_FAST:
+        case _LOAD_CONST_INLINE_BORROW:
+            return true;
+        default:
+            return false;
+    }
+}
+
+static inline bool
+op_is_load(int opcode)
+{
+    return (opcode == _LOAD_CONST_INLINE ||
+        opcode == _LOAD_CONST ||
+        opcode == LOAD_FAST ||
+        opcode == _LOAD_CONST_INLINE_BORROW);
+}
+
+static int
+peephole_optimizations(_PyUOpInstruction *buffer, int buffer_size)
+{
+    bool done = true;
+    for (int i = 0; i < buffer_size; i++) {
+        _PyUOpInstruction *curr = buffer + i;
+        int oparg = curr->oparg;
+        switch(curr->opcode) {
+            case _SHRINK_STACK: {
+                // If all that precedes a _SHRINK_STACK is a bunch of loads,
+                // then we can safely eliminate that without side effects.
+                int load_count = 0;
+                _PyUOpInstruction *back = curr-1;
+                while(op_is_zappable(back->opcode) &&
+                    load_count < oparg) {
```

Comment:
In order to avoid (theoretically possible) backing up before the start of the buffer, we should *definitely* check for `load_count < oparg` before calling `is_zappable()`, and we should *probably* also add a check that `back >= buffer`.
```suggestion
                while (back >= buffer && load_count < oparg && op_is_zappable(back->opcode)) {
```